### PR TITLE
Unified framework interface

### LINKED ISSUE #118: Torch test image "stops" training
If you classify an image while a torch model is still training, DIGITS marks the job as "Done", even though the training is still going on in the background.

### LINKED ISSUE #120: Torch doesn't do initial validation phase
Torch takes forever to actually output something on the loss/accuracy graph, and when something finally appears, it doesn't start at x=0. The way to fix this is to run a validation pass *before* beginning training.

Caffe by default does a pass through the validation data once before beginning to train the model ([see comments here](https://github.com/NVIDIA/caffe/blob/v0.11.0/src/caffe/proto/caffe.proto#L134-L136)). They do this partly to ensure memory availability (which shouldn't be an issue with Torch), but it also helps DIGITS to be able to display something quickly.

### LINKED ISSUE #121: Torch epoch rounding issue
Here is the output from torch training AlexNet:
```
2015-05-21 14:13:15 [INFO ] Training (epoch 0): loss = nan, lr = 0.01
2015-05-21 14:15:55 [INFO ] Training (epoch 0.01): loss = 0.99398752933095, lr = 0.01
2015-05-21 14:18:31 [INFO ] Training (epoch 0.01): loss = 1.0189151917673, lr = 0.01
2015-05-21 14:21:05 [INFO ] Training (epoch 0.02): loss = 0.9794481931978, lr = 0.01
2015-05-21 14:23:39 [INFO ] Training (epoch 0.02): loss = 1.0302608635775, lr = 0.01
2015-05-21 14:26:11 [INFO ] Training (epoch 0.02): loss = 0.98033467011087, lr = 0.01
```
DIGITS doesn't know what to do with two training loss values for epoch=0.01, so nothing gets displayed in the graph. We need to add more significant digits to the epoch so each update has a unique value.

### LINKED ISSUE #125: Torch branch can't display jobs from master branch
When I get the `torch` branch and try to look at a job trained with caffe on the `master` branch, I get this error:
```
BuildError
('serve_file', {}, None)

...
  File "/home/lyeager/digits/digits/templates/models/images/classification/show.html", line 30, in block "job_content"
    <dd><a href="{{url_for('serve_file', path=task.path(task.log_file, relative=True))}}">{{task.log_file}}</a></dd>
...
```

### LINKED ISSUE #128: Torch needs to adjust models for each dataset
When using Caffe, DIGITS adjusts the last layer of the network to match the number of classes in the dataset. This is pretty easy, just `layer[-1].num_output = num_classes`.

It's more complicated in Torch because the network is defined in a piece of lua source code instead of a text file. We need to find some kind of solution for this. Maybe after `main.lua` reads the model, it can edit that object and adjust the output of the last layer?

### PR DESCRIPTION
**Pull request by gheinrich**
Move (most of) the framework-specific code into unified interfaces: Framework and Train classes.

Squashed commit of the following:

commit 6e7922867f15f701f87a3cae77ef2421de464d96
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sun Aug 30 17:24:53 2015 +0200

    Framework-agnostic model files

commit fead6f44c87fdfcaba6906af14aa27203428b2cc
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sun Aug 30 10:59:02 2015 +0200

    Fix missing shuffle parameter

commit 1c0edd08cd26bd670e1c8d66a20cd71dce38f464
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sun Aug 30 10:45:19 2015 +0200

    Move framework-specific constants out of utils/constants.py

commit d710a2436e7464ed5b2048392daa3d79fc863acc
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sun Aug 30 10:15:27 2015 +0200

    Remove duplicate h5py requirement

commit 4e0c86b5656a85cd065a3de88a142d7001af1d9c
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sat Aug 29 22:42:38 2015 +0200

    Show framework ID on result page

commit d5b68300eaa228ee68866f0f1eb5c2fdc75bab15
Merge: d102d34 8a3d791
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sat Aug 29 22:16:37 2015 +0200

    Merge remote-tracking branch 'origin/master' into HEAD

    Conflicts:
    	docs/API.md
    	docs/FlaskRoutes.md

commit d102d341ac76f2d20be4d9d627b7208b90368103
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 28 16:34:50 2015 +0200

    Change comment style

commit 0ba079df9d3cf680959c1912c71c16f68014edfd
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 28 16:15:36 2015 +0200

    Report training loss on first iteration - closes #121

commit 11e0c24c4f52006c09c59aee340bd08fedf46263
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 28 12:12:22 2015 +0200

    Make Torch optional

    Also, don't show inference buttons if Torch framework is not
    present

commit 7ec0ded7b10112db6879b87699ccf7ec9c148431
Merge: 7483240 a75765a
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 28 11:07:15 2015 +0200

    Merge remote-tracking branch 'origin/master' into HEAD

commit 7483240510eb6a74745dddcc2e7ac4e529ec4ae4
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 28 11:00:02 2015 +0200

    Rework comments and some function names

commit 8d2ec42a20782c3552a6e308f5b81e60644c21d7
Merge: 048938c bbad11a
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Thu Aug 27 17:00:35 2015 +0200

    Merge remote-tracking branch 'origin/master' into HEAD

    Conflicts:
    	digits/model/images/classification/test_views.py
    	digits/templates/datasets/images/classification/show.html

commit 048938c06ca736efc3074e9fe003179232202def
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Thu Aug 27 16:31:24 2015 +0200

    Fix failed tests

commit 2482e677acf07133f560383a249857deea821556
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Thu Aug 27 15:43:49 2015 +0200

    Framework-agnostic network visualization and customization

commit 011e0f6a1592d0e2d172f20cbffd6edf2742e7b5
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Thu Aug 27 14:31:50 2015 +0200

    Isolation of framework changes for classify_one/many

commit 250ba0d59bc5adc0f20a8fef188ec68f3567fc78
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Mon Aug 24 14:12:08 2015 -0700

    Bugfix - wrong shape calculated for infer_many

    Conflicts:
    	digits/model/images/classification/test_views.py

commit 4c2edcc9554533da151dccbb34cb15a21fec7431
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Wed Aug 26 14:51:26 2015 +0200

    Framework-agnostic model file retrieval

commit ebe5d985557704a39feb06b5be97918a1227da6a
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Tue Aug 25 21:04:34 2015 +0200

    Partial isolation of framework-specific code

    Moving framework-specific code to a single generic Framework class.
    The class is derived into CaffeFramework and TorchFramework and can
    be instantiated an arbitrary number of times to support N frameworks
    in parallel.

    TODO: move framework config, classify_one/many, generic inference, etc. to Framework class.

commit 6e0568e4303bf5bd30e421fae554701d890d37d9
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Tue Aug 25 13:34:02 2015 +0200

    Restore option to shuffle data during Torch training

commit d116a89e44a467881b30f78ccc5217fb4c6576e3
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Tue Aug 25 12:08:45 2015 +0200

    Fix placement of model weights

commit e1282f9e6211ace9b691cc34539d60e8e4ad861f
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Mon Aug 24 14:18:56 2015 +0200

    Remove bogus <SCRIPT> tag

    This was preventing custom Torch networks from being entered.

commit 5ee40097115ef14288414b325cb579057befafb7
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Sun Aug 23 13:14:55 2015 +0200

    Don't show mean image during dataset creation

    To avoid broken-image icon being displayed as mean image file
    not yet created.

commit b55f1d0f876f3963906492aac4856be36508fea7
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 21 13:40:08 2015 +0200

    Add Torch model parameter count

commit cc7d27666b9370e8a513cea33aef5e7c746cbdbb
Merge: 65cb88a 3fca56d
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 21 13:24:34 2015 +0200

    Merge remote-tracking branch 'origin/master' into HEAD

    Conflicts:
    	digits/model/forms.py
    	digits/templates/models/images/classification/new.html
    	digits/templates/models/images/generic/new.html
    	docs/API.md
    	docs/FlaskRoutes.md

commit 65cb88a58f55ccaf987d65295e51297742fb312a
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Fri Aug 21 09:18:37 2015 +0200

    Add Torch visualization of weights and activations

commit b7bcb4c8a100370c71a1361cadecf1da4e248dfb
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Wed Aug 19 20:24:10 2015 +0200

    Use last snapshot if snapshot epoch unspecified

commit 0b94f02fbea437f8e0dc1be0e160012f7b8657bf
Merge: b82892b c47341a
Author: Greg Heinrich <gheinrich@nvidia.com>
Date:   Mon Aug 17 15:13:23 2015 +0200

    Merge remote-tracking branch 'origin/master' into HEAD

    Conflicts:
    	digits/model/forms.py
    	digits/model/images/classification/test_views.py
    	digits/model/images/classification/views.py
    	digits/model/tasks/torch_train.py
    	digits/templates/models/images/classification/new.html
    	digits/test_webapp.py
    	digits/utils/constants.py
    	docs/API.md
    	docs/FlaskRoutes.md
    	tools/test_create_db.py

commit b82892b5dee9006cf40744a72eb4074c6e1caa51
Merge: 0013ef6 4eb2f2c
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Mon Jun 22 13:09:26 2015 -0700

    Merge remote-tracking branch 'origin/master' into torch

    Conflicts:
    	digits/test_webapp.py

    Added some TODOs for @semisight

commit 0013ef69e10538fa2d001ad09b69f861236ba5e0
Author: Sravan <sravan2j@gmail.com>
Date:   Wed Jun 17 18:00:36 2015 -0700

    Include test functions for Torch framework

    1) Included test functions for Torch framework
    2) Coded model download feature for torch framework
    2) Fixed bugs related to Importing and learning rate calculation (in Torch)

commit 70f64f88d2aff04c9a2f909994ec98c92d7cb1d6
Merge: 1f849a6 6a11e66
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Mon Jun 15 17:10:08 2015 -0700

    Merge remote-tracking branch 'origin/master' into torch

    Conflicts:
    	.travis.yml
    	digits/model/forms.py
    	digits/model/images/classification/views.py
    	digits/model/views.py
    	digits/templates/home.html
    	docs/API.md
    	docs/FlaskRoutes.md
    	tools/test_create_db.py

commit 1f849a6be7f5f0706ab42d6f6309f669921a1d6c
Author: Sravan <sravan2j@gmail.com>
Date:   Mon Jun 15 15:49:01 2015 -0700

    Fix failed travis tests

commit 78b6ba465ddc4f319f7fd8dc9677a31809d4e303
Author: Sravan <sravan2j@gmail.com>
Date:   Mon Jun 15 14:26:44 2015 -0700

    Merge 'origin/master' into torch

commit a50ccfc023103256a106cedf165eeb55d0cd6d39
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Mon Jun 15 14:02:05 2015 -0700

    Add docs for installing LMDB from source

    Figured it out for Travis

commit 5ce8c98f85451757b2e4f31e872fb5c866c90616
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Mon Jun 15 13:46:15 2015 -0700

    Fix upgrade path for CaffeTrainTask

commit fc55d194e6ee8b6b1836eda628a27dd6b17cd65e
Author: Sravan <sravan2j@gmail.com>
Date:   Thu Jun 11 10:39:02 2015 -0700

    display framework beside jobs in "Completed" panel

    Display framework name just beside jobs in "Completed" panel in the home
    screen

commit fef1c8420a2ccbf48f3ae6e163b5bf3a2ae5c6d1
Merge: 6578bd1 eba9a2c
Author: Sravan <sravan2j@gmail.com>
Date:   Tue Jun 9 16:20:19 2015 -0700

    Merge 'origin/master' into torch and fix #121

    1) merge 'origin/master' into torch and resolve below conflicts
    Conflicts:
    	digits/config.py
    	digits/model/images/classification/views.py
    2) fix val_interval value truncation mentioned in issue #121 comments.
    3) user can provide seed value to random function
    4) included additional solver types to torch.

commit 6578bd1fc32f1897a87a98987294c6a6e7260e49
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 29 10:53:40 2015 -0700

    fix #125

    1) Update caffe jobs with pickle version 1 and 2, to use log_file attribute
    instead of caffe_log_file

commit 51314c1ac1e1f4f0f023818e0784d8036c2881c6
Merge: 028baf1 f968eb4
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Thu May 28 10:59:44 2015 -0700

    Merge branch 'origin/master' into torch

    Conflicts:
    	digits/config.py
    	digits/model/forms.py
    	digits/model/images/classification/views.py
    	digits/model/tasks/torch_train.py

commit 028baf172791065c6f9c85a997c88ee2ff7f5f99
Author: Sravan <sravan2j@gmail.com>
Date:   Wed May 27 18:42:47 2015 -0700

    fix #128

    1) changes to adjust model for subset of dataset
    2) code changes to process train output properly and properly display error message

commit 2eb0b52537b8bc673b5bba2382a3fdd20145b41a
Author: Sravan <sravan2j@gmail.com>
Date:   Mon May 25 03:06:08 2015 -0700

    closes #120, #121, 122

    1) This code changes fixes all the three issues - #120, #121 and #122
    2) During training, after each batch is processed, batchsize is added to processed images count. Sometimes the final batch size may be less than required batch size. In this scenario min() function is used to get the details of exact images count that are processed till that time.

commit a973d492f0cc0fefbfded19395197f10011ed55e
Merge: d25a686 a350084
Author: Sravan <sravan2j@gmail.com>
Date:   Thu May 21 19:18:50 2015 -0700

    Merge branch 'torch' of https://github.com/NVIDIA/DIGITS into torch

commit d25a68695e379da6969fd0535bbb4fe0f124dc17
Author: Sravan <sravan2j@gmail.com>
Date:   Thu May 21 19:14:02 2015 -0700

    fix #118

    1) created new exception and used those exceptions to notify the user
    about the issues faced while testing.
    2) corrected logic to find "unrecognized messages" from train and testing output

commit a350084f891b6f7132cf0cfbf0a82b68b34a5baf
Author: Luke Yeager <lyeager@nvidia.com>
Date:   Thu May 21 11:32:46 2015 -0700

    More changes to torch install instructions

    Markup changes
    Remove TORCH_INSTALL environment variable suggestion - the torch
    default is to install for the whole system
    Add image to the install list because it needs to be upgraded
    Move ccn2 to the bottom of the install list since it takes so long

commit 333dff339ca177d3120fe23d476cbfe512069b76
Author: Luke Yeager <lukeyeager@users.noreply.github.com>
Date:   Wed May 20 15:15:23 2015 -0700

    Update InstallTorch.md

    More spelling errors

commit e1b8694af60d393810f2ffbc2a5d5b676dabf879
Author: Luke Yeager <lukeyeager@users.noreply.github.com>
Date:   Wed May 20 15:14:49 2015 -0700

    Update InstallTorch.md

    Spelling error

commit 1d562261cd3b999c19d69633cde6becf966fd7e6
Merge: f5d350c c3833cd
Author: Sravan <sravan2j@gmail.com>
Date:   Tue May 19 17:27:44 2015 -0700

    Merge branch 'lukeyeager-torch' into torch

commit c3833cdde6db0881237e33799ed5f3ed9264a01b
Author: Sravan <sravan2j@gmail.com>
Date:   Tue May 19 17:24:19 2015 -0700

    rename torch install file

    1) Renamed TorchInstall.md file to InstallTorch.md
    2) changed the installation instructions for lmdb and lightningdbm

commit ddcae977d2c075b057c69e60b45022474e91bef1
Author: Luke Yeager <lukeyeager@users.noreply.github.com>
Date:   Mon May 18 18:13:14 2015 -0700

    Simplify torch installation instructions

    Fix markup and add envvar export
    Remove a bunch of unnecessary explanations
    Move LMDB instructions to the bottom

commit f5d350cae19b5cda0167eac17652742c7ca0d0e9
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 15 17:51:23 2015 -0700

    removed unnecessary assignment

commit 503096cf3ef9f6fc037863473897d3406bb413c0
Merge: 2379a79 a5311a0
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 15 15:44:39 2015 -0700

    Merge remote-tracking branch 'origin/master' into torch

commit 2379a796009506c3642c3c8a9506aac5cefec488
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 15 15:43:31 2015 -0700

    changes to find torch binary correctly

commit 2f6589053f99d9220a263e0a6ba8fad7a1f2daf7
Merge: 1c6000e 1c8a0d7
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 15 13:39:51 2015 -0700

    Merge remote-tracking branch 'origin/master' into torch

    Conflicts:
    	digits/config.py
    	digits/model/images/classification/test_views.py

commit 1c6000e196d73c5a0dcc3bec87c332ed9f0cba93
Author: Sravan <sravan2j@gmail.com>
Date:   Thu May 14 16:40:04 2015 -0700

    support PNG encoding

    1) replaced decompressJPG() with decompress() to support both JPG and
    PNG format
    2) corrected typos

commit 77009ec2ce694c661252a12f891d874b51c5d3e5
Author: Sravan <sravan2j@gmail.com>
Date:   Wed May 13 15:29:18 2015 -0700

    fine tuning changes

    1) code changes to copy weights from pretrained model for fine tuning
    2) Also completed modules for "crash from recovery" feature
    3) code clean up

commit aac6e421d40155562b6a107e0e79d5bb04db8604
Author: Sravan <sravan2j@gmail.com>
Date:   Tue May 12 11:31:49 2015 -0700

    Change lua-pb installation instruction

    Changed lua-pb installation instruction to point to sravan2j github
    repository to avoid issues while installing.
    Included installation instruction for inn module

commit b05c91b8f893f079095ef56545a071852a185437
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 8 11:27:37 2015 -0700

    fixed bugs and did code clean up

    1) During shuffle, images labels are not correctly stored. Image labels
    are in Caffe format i.e., label values starts from 0. But in Torch,
    Tensors are indexed from 1. So, changed the code accordingly.
    2) Did changes to TorchInstall.md file to correctly parse the values in
    between "<" and ">" and also corrected some installation instructions
    3) Did code clean up like removing unnecessary code.

commit 25fe83be06f845bd582a0a10b7d745265c0c9997
Author: Sravan <sravan2j@gmail.com>
Date:   Tue May 5 18:31:44 2015 -0700

    Seperated textareas for caffe and torch

    Seperate textarea for each framework is declared in custom tab to
    preserve the network for caffe and torch at the same time.

commit 174d287ac8dce5d219cc976e4a99b54f29138b62
Author: Sravan <sravan2j@gmail.com>
Date:   Tue May 5 13:55:17 2015 -0700

    did changes as per Soumith suggestions

commit cb1e9a5011d24cccb2a8eebfd5dec5d86f64f42e
Author: Sravan <sravan2j@gmail.com>
Date:   Fri May 1 18:37:50 2015 -0700

    Fixed bugs and added shuffle option to UI

    1) Increased Tensor size by 1 than required, so that null terminator
    i.e., '\0', also gets copied to Tensor. This fixed the memory corruption
    bug in Torch code.
    2) Sometimes execution of torch job from "previous tab" throws a "caffe
    network" related error. This is because views.py is by mistakenly
    detecting torch job as caffe job. Did code changes to retrieve the
    framework details from the selected previous job itself instead of using
    form.framework.data
    3) Added "shuffle" option to UI. This option is shown only for torch
    jobs.

commit aa972c5f1a71939029e6076bb9d75b3ec20c83f6
Author: Sravan <sravan2j@gmail.com>
Date:   Thu Apr 30 15:45:24 2015 -0700

    Fixed the bug related to JPG encoded images

    encoded JPG images have more byte size than the total size (i.e., channel x width x height). So increased the tensor size to 5 times more than than original size. Probably 2 times should be effecient, but just to be safe side allocated 5 times more size.

commit e4fcc5db5d3add1ce349599634f623e2741d4b53
Author: Sravan <sravan2j@gmail.com>
Date:   Wed Apr 29 13:37:04 2015 -0700

    correction to validation message

commit 38bde676a9bd420cf38ce1c90911c07c080bcba6
Author: Sravan <sravan2j@gmail.com>
Date:   Mon Apr 27 17:11:49 2015 -0700

    Included Torch installation instructions

commit bef0d4b893b04e7315dc66a21717daa82fdb04b8
Author: Sravan <sravan2j@gmail.com>
Date:   Fri Apr 24 16:42:56 2015 -0700

    included code to test multiple images once

commit 02b0f585d4019fa5032f07ede41dc80214ecb576
Author: Sravan <sravan2j@gmail.com>
Date:   Mon Apr 20 17:56:18 2015 -0700

    added torch option to custom tab and previous tab

commit 0ef16c8bf92c348a978c2d35149b2eab53454b72
Author: Sravan <sravan2j@gmail.com>
Date:   Mon Apr 20 15:43:21 2015 -0700

    completed logic to test multiple images

commit 0810df478e4445a325130b1b396cc6f381fa139e
Author: Sravan <sravan2j@gmail.com>
Date:   Fri Apr 17 14:51:33 2015 -0700

    Added test a single image feature in Torch. Also did code clean up

commit 9ea225cccd5ea1d838568ae891df002e89d4ffd2
Author: Sravan <sravan2j@gmail.com>
Date:   Wed Apr 8 11:08:54 2015 -0700

    changes to model code write logic of custom tab

commit ee9e3c124617cc067b54d4335083b126aa9b7400
Author: Sravan <sravan2j@gmail.com>
Date:   Thu Apr 2 17:13:33 2015 -0700

    added changes to open lmdb in read-only mode

    Changes to Make LMDB shareable between jobs.
    Corrected code to display valuation accuray and loss values on the graph

commit 41ab1b8f8efa6af5fbb1b6b8505cd5aeb2a12639
Author: Sravan <sravan2j@gmail.com>
Date:   Wed Apr 1 18:43:12 2015 -0700

    torch changes to take snapshots at given interval

commit 75d9a960150959adc7b47b09359cb31cc1acaf69
Author: Sravan <sravan2j@gmail.com>
Date:   Wed Apr 8 13:29:12 2015 -0700

    Included torch7 into DIGITS

    This adds just a basic Torch7 functionality. Below are some pending tasks:
    1) Parallelize Data loading and Training steps to improve performance
    2) Torch code is giving memory exception when shuffle parameter is set
    to "yes". Probably the error deals with the logic for random access of
    data from LMDB
    3) Make use of pretrained model i.e., to continue training from the pretrained model
    4) image.decompressJPG() function is giving an error for the grey scale images (eg: MNIST dataset) that is encoded using Caffe code. For the color images, decompress works fine.
    5) Visualize Torch Model

Remove references to Torch

### CODE REVIEW

File: digits/frameworks/__init__.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Copyright (c) 2014-2015, NVIDIA CORPORATION.  All rights reserved.
```
Comment (lukeyeager) [inline digits/frameworks/__init__.py]: This should be just `2015`.

File: digits/frameworks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+def get_frameworks():
+    """
+    return list of all available framework instances
+    there may be more than one instance per framework class
+    """
+    frameworks = [caffe]
+    return frameworks
+
+def get_framework_by_id(id):
```
Comment (lukeyeager) [inline digits/frameworks/__init__.py]: It's bad form to use `id` as a variable name since it's also a python builtin function.

https://docs.python.org/2/library/functions.html#id

File: digits/frameworks/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from digits.utils import subclass, override
```
Comment (lukeyeager) [inline digits/frameworks/exceptions.py]: Needs a copyright statement.

File: digits/frameworks/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+@subclass
+class BadNetworkException(Exception):
```
Comment (lukeyeager) [inline digits/frameworks/exceptions.py]: Having custom exceptions for the Frameworks module is awesome!

Can we follow [this convention](https://docs.python.org/2/tutorial/errors.html#user-defined-exceptions) and create two classes with these names?
```python
class Error(Exception):
    pass
class BadNetworkError(Error):
    pass
```
I know we haven't been consistent with this before, but I'd like to move in that direction. See here:
https://github.com/NVIDIA/DIGITS/blob/4121e3a1c0/tools/create_db.py#L41-L58

File: digits/model/images/classification/test_views.py [hunk 5]
```diff
@@ -70,6 +70,10 @@ class BaseViewsTest(digits.test_views.BaseViewsTest):
 """
 
     @classmethod
+    def setUpClass(cls):
+        super(BaseViewsTest, cls).setUpClass()
```
Comment (lukeyeager) [inline digits/model/images/classification/test_views.py]: Isn't this done automatically if you don't explicitly declare a `setUpClass()` method? Not a big deal - I'm just curious.

File: digits/task.py [hunk 6]
```diff
@@ -239,7 +239,10 @@ def run(self, resources):
             if self.exception is None:
                 self.exception = 'error code %d' % p.returncode
                 if unrecognized_output:
-                    self.traceback = '\n'.join(unrecognized_output)
+                    if self.traceback is None:
+                        self.traceback = '\n'.join(unrecognized_output)
+                    else:
+                        self.traceback = self.traceback + ('\n'.join(unrecognized_output))
```
Comment (lukeyeager) [inline digits/task.py]: Is this related to your changes or just a bugfix?

File: digits/utils/errors.py [hunk 7]
```diff
@@ -17,3 +17,9 @@ class LoadImageError(DigitsError):
     Errors that occur while loading an image
     """
     pass
+
+class TestError(DigitsError):
+    """
+    Errors that occur while loading an image
+    """
+    pass
```
Comment (lukeyeager) [inline digits/utils/errors.py]: I'd actually like to remove this file if possible. Is there a more logical place to put this class?

Also, it looks like you copied the docstring without editing it.

File: tools/resize_image.py [hunk 8]
```diff
@@ -99,7 +99,6 @@ def validate_range(number, min_value=None, max_value=None, allow_none=False):
             )
 
     args = vars(parser.parse_args())
-
```
Comment (lukeyeager) [inline tools/resize_image.py]: Can you remove this whitespace change?

**Comment by lukeyeager** [comment]
Haha wow that's a long commit message! Is it feasible to squash the commits that Sravan did in a separate commit to preserve author information in the git log? I think it's worth it even if it adds more than one extra commit.

This looks great, thanks for the great work! I added some comments above - they're all pretty small changes.

File: digits/templates/models/images/classification/new.html [hunk 9]
```diff
@@ -340,7 +360,26 @@
                 <li><a href="#network-tab-custom" data-toggle="tab">Custom Network</a></li>
             </ul>
             <div class="tab-content well">
+                <div style="display:none;">{{form.framework(class='form-control')}}</div>
                 <div id="network-tab-standard" class="tab-pane active">
+                    <ul class="nav nav-tabs" id="stdnetRole">
+                        {% for fw in frameworks %}
+                        <li id="{{ fw.get_id() }}"><a href="#">{{ fw.get_name() }}</a></li>
```
Comment (lukeyeager) [inline digits/templates/models/images/classification/new.html]: The `href="#"` doesn't really make sense here (and anchors to the top of the page if you click on it). I think you want a structure kinda like this:
```html
<div id="network-tab-standard" class="tab-pane active">
  <ul class="nav nav-tabs" id="stdnetRole">
    <li class="active"><a href="#standard-networks-caffe-tab">Caffe</a></li>
  </ul>
  <div id="standard-networks-caffe-tab" class="tab-pane active">
    Content here.
  </div>
</div>
```

File: digits/model/images/classification/test_views.py (hunk 5 above)
Comment (gheinrich) [inline digits/model/images/classification/test_views.py]: Yes indeed, it is actually a left over from the torch branch where special action was taken in case this was a torch test. I can remove the method completely for the time being. Thanks.

File: digits/task.py (hunk 6 above)
Comment (gheinrich) [inline digits/task.py]: This is from Sravan's commit 028baf172791065c6f9c85a997c88ee2ff7f5f99:
```
 fix #128

 1) changes to adjust model for subset of dataset
 2) code changes to process train output properly and properly display error message
```

Comment (lukeyeager) [inline digits/task.py]: Ah ok.

**Comment by lukeyeager** [comment]
LGTM, thanks!
